### PR TITLE
Add log warn to OAuth2::Errors

### PR DESCRIPTION
**Pull request by Thource**
### Jira link

P4-2128

### What?

I have added/removed/altered:

- [x] Added logging to OAuth2::Errors raised through NomisClient
- [x] Added unit tests to cover OAuth2::Error logging and Faraday error logging

### Why?

I am doing this because:

- It will make it more clear where the error is coming from

### CODE REVIEW

**Review by willfish — APPROVED** [review]
Seems fine to me

**Comment by willfish** [comment]
Congrats on first PR 🎉

File: app/lib/nomis_client/base.rb [hunk 1]
```diff
@@ -29,6 +29,9 @@ def token_request(method, path, params)
       rescue Faraday::ConnectionFailed, Faraday::TimeoutError => e
         Rails.logger.warn "Nomis Connection Error: #{e.message}"
         raise e
+      rescue OAuth2::Error => e
```
Comment (cwrw) [inline app/lib/nomis_client/base.rb:32]: This seems similar to the above rescue block, maybe we can combine them?

Comment (Nimphal) [reply app/lib/nomis_client/base.rb:32]: The message is specific to the errors, I can't think of a neat way to combine them and retain the message granularity, but keen to learn!

File: spec/lib/nomis_client/base_spec.rb [hunk 2]
```diff
@@ -306,4 +306,40 @@
...
+          allow(Rails).to receive(:logger).and_return(instance_spy('logger'))
+        end
+
+        %i[post put get].each do |method|
+          context "when the request type is #{method}" do
+            before do
+              allow(token).to receive(method).and_raise(oauth2_error)
+            end
+
+            it 'logs additional info, describing what the error is' do
```
Comment (cwrw) [inline spec/lib/nomis_client/base_spec.rb:336]: Generally I don't think we should be testing logging in the app as its usually superfluous, but happy to hear other's thoughts around this.

File: app/lib/nomis_client/base.rb (hunk 1 above)
Comment (Thource) [reply app/lib/nomis_client/base.rb:32]: The only thing I can think of would be something like

```ruby
Rails.logger.warn "Nomis #{e.is_a? OAuth2::Error ? 'Oauth Client' : 'Connection'} Error: #{e.message}"
```

Comment (cwrw) [reply app/lib/nomis_client/base.rb:32]: That's fair, I was hoping the error message would give us the details of the error itself, but happy to keep as is

File: spec/lib/nomis_client/base_spec.rb (hunk 2 above)
Comment (willfish) [reply spec/lib/nomis_client/base_spec.rb:336]: I don't think I agree. It's a test that tells us we're logging errors (which is important) and doesn't have much cost for the value it gives us if we no longer log these due to a regression. If we were logging other stuff like random unimportant messages I'd argue that this shouldn't be tested. But the value aspect makes me think this is fine.

**Pull request merged**